streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryWindowStore.java

```diff
@@ -279,39 +275,74 @@ private static Bytes getKey(final Bytes keyBytes) {
 
     }
 
-    private abstract class InMemoryWindowStoreIteratorWrapper implements Comparable<InMemoryWindowStoreIteratorWrapper> {
+    private WrappedInMemoryWindowStoreIterator registerNewWindowStoreIterator(final Bytes key,
+                                                                              final Iterator<Map.Entry<Long, ConcurrentNavigableMap<Bytes, byte[]>>> segmentIterator) {
+        final Bytes keyFrom = retainDuplicates ? wrapForDups(key, 0) : key;
+        final Bytes keyTo = retainDuplicates ? wrapForDups(key, Integer.MAX_VALUE) : key;
+
+        final WrappedInMemoryWindowStoreIterator iterator =
+            new WrappedInMemoryWindowStoreIterator(keyFrom, keyTo, segmentIterator, openIterators::remove);
+
+        openIterators.add(iterator);
+        return iterator;
+    }
+
+    private WrappedWindowedKeyValueIterator registerNewWindowedKeyValueIterator(Bytes keyFrom,
+                                                                                   Bytes keyTo,
+                                                                                   final Iterator<Map.Entry<Long, ConcurrentNavigableMap<Bytes, byte[]>>> segmentIterator) {
+        if (retainDuplicates && !(keyFrom == null || keyTo == null)) {
+            keyFrom = wrapForDups(keyFrom, 0);
+            keyTo = wrapForDups(keyTo, Integer.MAX_VALUE);
+        }
+
+        final WrappedWindowedKeyValueIterator iterator =
+            new WrappedWindowedKeyValueIterator(keyFrom,
+                                                keyTo,
+                                                segmentIterator,
+                                                openIterators::remove,
+                                                retainDuplicates,
+                                                windowSize);
+        openIterators.add(iterator);
+        return iterator;
+    }
+
+
+    interface ClosingCallback {
+        void deregisterIterator(final InMemoryWindowStoreIteratorWrapper iterator);
+    }
+
+    private static abstract class InMemoryWindowStoreIteratorWrapper {
 
         private Iterator<Map.Entry<Long, ConcurrentNavigableMap<Bytes, byte[]>>> segmentIterator;
         private Iterator<Map.Entry<Bytes, byte[]>> recordIterator;
         private KeyValue<Bytes, byte[]> next;
         private long currentTime;
 
         private final boolean allKeys;
-        private Bytes keyFrom;
-        private Bytes keyTo;
+        private final Bytes keyFrom;
+        private final Bytes keyTo;
+        private final ClosingCallback callback;
 
         // Default constructor sets up a dummy iterator when no results are returned (eg entire fetch range is expired)
         InMemoryWindowStoreIteratorWrapper() {
-            this.allKeys = false;
+            allKeys = false;
             recordIterator = null;
+            keyFrom = null;
+            keyTo = null;
+            callback = null;
```

Comment:
The field members here as well as all other object field members are all initialized to `null`. However, I would either explicitly initialize all field members with the default values if the value is not specified by a parameter as done here for `recordIterator` or rely on the default initialization for all. What is the project policy on this?

Reply:
Yeah, this is really just supposed to be a dummy iterator (which is effectively the case when recordIterator = null) but I had to also initialize all members that were final. I think it's better to just generate a dummy iterator with a static method instead of this dummy constructor so I'll probably remove this